frontends/p4/functionsInlining.cpp

```diff
@@ -97,20 +124,46 @@ void FunctionsInliner::dumpReplacementMap() const {
     auto replMap = getReplacementMap();
     LOG2("Replacement map contents");
     if (replMap == nullptr) return;
-    for (auto it : *replMap) LOG2("\t" << it.first << " with " << it.second);
+    for (auto it : *replMap)
+        LOG2("\t" << it.first << " with " << it.second.first << " via " << it.second.second);
 }
 
 const IR::Node *FunctionsInliner::preorder(IR::AssignmentStatement *statement) {
     auto orig = getOriginal<IR::AssignmentStatement>();
     LOG2("Visiting " << dbp(orig));
+
     auto replMap = getReplacementMap();
     if (replMap == nullptr) return statement;
 
-    auto callee = get(*replMap, orig);
+    auto [callee, callExpr] = get(*replMap, orig);
     if (callee == nullptr) return statement;
-    auto call = statement->right->to<IR::MethodCallExpression>();
-    BUG_CHECK(call != nullptr, "%1%: expected a method call", statement->right);
-    return inlineBefore(callee, call, statement);
+    BUG_CHECK(callExpr != nullptr, "%1%: expected a method call", statement->right);
+    return inlineBefore(callee, callExpr, statement);
+}
+
+const IR::Node *FunctionsInliner::preorder(IR::Declaration_Variable *decl) {
+    auto orig = getOriginal<IR::Declaration_Variable>();
+    LOG2("Visiting " << dbp(orig));
+
+    auto replMap = getReplacementMap();
+    if (replMap == nullptr) return decl;
+
+    auto [callee, callExpr] = get(*replMap, orig);
+    if (callee == nullptr) return decl;
+
+    BUG_CHECK(callExpr != nullptr, "%1%: expected a method call to initialize", decl);
+
+    // To simplify the implementation we are cloning the variable w/o
+    // initializer and pretending that there is explicit assignment there.
+    auto *newDecl =
+        new IR::Declaration_Variable(decl->srcInfo, decl->getName(), decl->annotations, decl->type);
+    LOG3("Cloned " << dbp(newDecl));
+
+    auto *assign = new IR::AssignmentStatement(
+        decl->srcInfo, new IR::PathExpression(newDecl->getName()), decl->initializer);
+    auto *res = inlineBefore(callee, callExpr, assign);
+
+    return new IR::Vector<IR::StatOrDecl>{newDecl, res};
```

Comment:
I am assuming this requires a block expression to be legal in the calling context.
is this always true?

Reply:
I might be missing something, but it seems to be so:
  - `DiscoverFunctionsInlining` would collect `Declaration_Variable`'s that have `MethodCallExpression` inside the initializer and only those.
  - In fact, at this time the only cases for such `Declaration_Variable`'s are the ones that were created by `functionInliner` by itself on the earlier iteration (for function arguments), everything else is already replaced by initializer-less ones, plus explicit `AssignmentStatement`.

Reply:
the only thing I can think of are constant declarations.
If a constant initializer can involve action calls, it may be a problem.

Reply:
I have difficulty imagining how a constant initializer could involve an action call, since actions calls are not expressions that return values.  Thus I believe they cannot be part of an expression, only a full statement by themselves.

I constant initializer could involve function calls, of course.  Not sure if that is relevant here.

Reply:
constants are `Declaration_Constant`, so they are excluded here. And per spec:
```
Variables declarations can appear in the following locations within a P4 program:

 * In a block statement,
 * In a parser state,
 * In an action body,
 * In a control block's apply sub-block,
 * In the list of local declarations in a parser, and
 * In the list of local declarations in a control.
```

So, I think block expression is legal in all these places.

Reply:
the local declarations may not hold statements I think

Reply:
> the local declarations may not hold statements I think

Oh, right, but can they have initializers there?

Reply:
Initially, yes, but the compiler desugars that quickly and I think it maintains the invariant that it never adds there intializers.

Reply:
Maybe then it would be better not to create `Declaration_Variable` with initializers during inlining at https://github.com/p4lang/p4c/blob/5c1e37f04f101482b1793a1926ddec52e5d8252c/frontends/p4/functionsInlining.cpp#L239 ?

So, instead, emit explicit `AssignmentStatement`? This might be better if further passes not expect initializers at all?

Reply:
For now I think you can leave it as it is. If somebody complains you have a hypothesis.
Or you can try to write a test to see if you can force declarations in this position. Maybe it's impossible. Check the shape of the program before this pass is run.

Reply:
I mean, if there is an invariant that variable initializers are already de-sugared at this stage, then we should not create new ones, right? This would actually make `Declaration_Variable` case here unused as everything will be handled by the existing `AssignmentStatement` handle. This probably will be cleaner solution, let me try this.

Reply:
I think that the MoveDeclarations is the pass which does this.

Reply:
Yes, MoveDeclarations does this.

Reply:
Ok, I went ahead and just create uninitialized variable there. It seems that existing `AssignmentStatement` case perfectly handles it, no need for something additional.